### ISSUE TITLE
LULC: Plotting

### ISSUE THREAD

**Issue by georgemccabe** [NCAR/i-wrf#70]
Sub-issue of #31 - LULC use case

## Describe the Task ##
Somewhat overlapping with the #69, develop scripts that can make relevant plots of the WRF output and METplus statistics, and then be plugged into the verification/visualization container. We might be able to leverage something from Xin's plotting script for his paper, but likely need to write something entirely new. Jared can update/adapt the generic WRF output plotting script he used for the Matthew case and simply update it with some useful model fields to plot for this use case (Xin, let Jared know what wrfout fields you'd like to have users generate plots of). That script (or a complementary one) could also be used to generate any other custom plots that are put on a map of the domain, if we want to keep visual continuity.

### Time Estimate ###
*Estimate the amount of work required here.*
*Issues should represent approximately 1 to 3 days of work.*

### Sub-Issues ###
Consider breaking the task down into sub-issues.
- [ ] *Add a checkbox for each sub-issue here.*

### Relevant Deadlines ###
0.3 (December 6, 2024)

### Funding Source ###
I-WRF

### Assignee ###
- [ ] Select **engineer(s)** or **no engineer** required
- [ ] Select **scientist(s)** or **no scientist** required

### Labels ###
- [ ] Select **component(s)**
- [ ] Select **priority**
- [ ] Select **requestor(s)**

### Projects and Milestone ###
- [ ] Select **Repository** and/or **Organization** level **Project(s)** or add **alert: NEED PROJECT ASSIGNMENT** label
- [ ] Select **Milestone** as the next official version or **Future Versions**

## Task Checklist ##
- [x] Complete the issue definition above, including the **Time Estimate** and **Funding Source**.
- [x] Fork this repository or create a branch of **develop**.
Branch name: `feature_<Issue Number>_<Description>`
- [ ] Complete the development and test your changes.
- [ ] Add/update log messages for easier debugging.
- [ ] Add/update unit tests.
- [ ] Add/update documentation.
- [ ] Push local changes to GitHub.
- [ ] Submit a pull request to merge into **develop**.
Pull request: `feature <Issue Number> <Description>`
- [ ] Define the pull request metadata, as permissions allow.
Select: **Reviewer(s)** and **Development** issues
Select: **Repository** level development cycle **Project** for the next official release
Select: **Milestone** as the next official version
- [ ] Iterate until the reviewer(s) accept and merge your changes.
- [ ] Delete your fork or branch.
- [ ] Close this issue.

**Comment by jaredalee** [comment]
I updated the task description above to reflect that we need to write a new script to plot output from METplus statistics (likely METplotpy) rather than directly using/adapting Xin's plotting script from his paper (as he didn't use METplus). Therefore, someone other than Xin should be co-assigned to this task as well, as he won't be primarily responsible for this. @georgemccabe or @briannen, which one of you should be co-assigned this instead of @levinzx? I'll remain a co-assignee, as I'll develop the WRF plotting script for this task.

**Comment by georgemccabe** [comment]
@briannen, do you have experience running METplotpy to generate these plots?

**Comment by briannen** [comment]
@georgemccabe @jaredalee I have some experience with METplotpy. Not much, but probably enough to create a few plots.

**Comment by georgemccabe** [comment]
@briannen, OK, great. I'll assign you to this issue. I have limited experience running METplotpy but I could take a look with you if you run into any issues. Thanks!

**Comment by jaredalee** [comment]
I have finished developing a WRF/NEXRAD plotting script that produces multi-panel plots per valid time, with two panels for fields from the two WRF experiments (control and DFW 4x), and for column-max radar reflectivity and accumulated rainfall, a third column plotting the NEXRAD observations. Example plots are attached.

I created a new branch (`feature_70_lulc_plotting`) and added the main Python script and a script with helper functions to use_cases/Land_Use_Land_Cover/Visualization. A few notes:

- These scripts currently are configured with defaults to read from already-existing WRF & NEXRAD data files on GLADE (Derecho/Casper).
- These scripts need to be tested and probably modified slightly in order to work within the context of reading WRF output for the control simulation generated from within the WRF container and staged WRF output for the DFW 4x simulation and NEXRAD observation files.
- These scripts require python-3.11.x or greater to run.

![Image](https://github.com/user-attachments/assets/d05aa6d0-2b16-44fd-9e33-d84ad99c9540)
![Image](https://github.com/user-attachments/assets/4bb3ea7c-3de9-4f3b-a3f8-2aa69bdc4bd8)

@briannen, could you please take a look at this and also develop some METplotpy scripts to complement these WRF/NEXRAD map plotting scripts?

**Comment by jaredalee** [comment]
@georgemccabe, I just realized that the WRF plotting script that's part of `feature_70_lulc_plotting` that I mentioned above, but which was never merged into `main`, has not yet been incorporated into the METplus container for the LULC case (it's in use_cases/Land_Use_Land_Cover/Visualization in the feature_70 branch). To read from the WRF and NEXRAD data volumes mounted to the container, and to write image files to an output directory within/visible to the container, see lines 918–929 of plot_wrf_lulc.py; either the default values would need to be changed, or correct file paths can be provided to the script as arguments. 

The directions in the user guide Sec. 6.2.4.1 Visualize the Results, simply says:

> In the near future, this exercise will be extended to include instructions to visualize the model output directly.

I hadn't noticed that before, but that's what alerted me to these WRF plots being missing from the instructions. I just went through the METplus container instructions step-by-step to verify that those plots are not currently being created.

@j-opatz, this is a change that I'd also like you to propagate forward as you copy the METplus container instructions over to the Jetstream2 and Red Cloud sections, if it gets made to the Casper instructions before the end of next week (otherwise @georgemccabe or someone else can take care of it).

**Comment by georgemccabe** [comment]
@jaredalee , thanks for catching this. I have an open [branch](https://github.com/NCAR/i-wrf/compare/main...feature_101_lulc_metplus?expand=1) with changes to add the METplus/plotting portion to the Jetstream2 and Redcloud LULC instructions. @zw427 is testing those additions on Jetstream2 and is going to add instructions to visualize the results on Redcloud.

I'll take a look at your branch to get the changes to run the LULC WRF plotting script into main. That portion is run via the METplus call, so the instructions won't change to add it. We could add instructions to tell users how to check that the WRF plots are generated properly by listing directory contents.

**Comment by jaredalee** [comment]
Sounds good, @georgemccabe. And maybe another enhancement that could be made to the RTD instructions would be to include an image or two of what users should see if they open one of those .png files, to give them confidence that they're seeing what they should see and that the container behaved as expected.

**Comment by jaredalee** [comment]
@georgemccabe Here's where the output for the two WRF experiments (ctl and dfw4x) are located on GLADE:

```
/glade/campaign/ral/wsap/i-wrf/science_case2_LULC_modification/wrfout/0703dfw/morr_nc300/ctl
/glade/campaign/ral/wsap/i-wrf/science_case2_LULC_modification/wrfout/0703dfw/morr_nc300/dfw4x
```

The output from dfw4x is what's already in a dockerhub data volume, but as we discussed today there were some inconsistencies in how the files were named between the data volume and if people would run the I-WRF container themselves. I like your idea, George, of creating two new data volumes for these two experiments and ensuring that the file names are consistently handled here and in the scripts that read these files.

**Comment by georgemccabe** [comment]
I created new data volumes for dfw4x and ctl WRF simulation data which use the naming conventions that are created by running WRF. I updated the METplus config files and NSF NCAR HPC instructions to and confirmed that it runs successfully using the new dfw4x file names.

I will need to update the instructions on the other platforms to ensure the PCPCombine input URL is correct -- this involves mounting the WRF directories to the appropriate directories in the container.

Once that is complete, I will add instructions to run Jared's WRF plotting script on the 2 simulations.
